### PR TITLE
Preparation for versioned application services

### PR DESCRIPTION
**Pull request by Whathecode**
This PR prepares for setting up a versioning strategy for application service requests and integration events.

For now, it sets up unit tests per application service which replay the existing application service unit tests and capture all requests, events, and responses. This whole setup could already be reviewed at this stage.

It also already outputs these logged requests to the build folder.

The next step is writing unit tests that try to deserialize the previously generated test sources for specific _API versions_, and includes a migration proxy. If the request succeeds after migrating, it validates the migration proxy. Some more work will be needed to validate migration of responses and events, which will likely require injecting custom `DateTime.now()` and `UUID.random` generators so they are constant.

### CODE REVIEW

File: carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/infrastructure/DataStreamServiceLog.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import dk.cachet.carp.data.application.DataStreamId
+import dk.cachet.carp.data.application.DataStreamService
+import dk.cachet.carp.data.application.DataStreamsConfiguration
+
+
+/**
+ * A proxy for a data stream [service] which notifies of incoming requests and responses through [log]
+ * and keeps a history of requests in [loggedRequests].
+ */
+class DataStreamServiceLog(
```
Comment (ltj) [inline carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/infrastructure/DataStreamServiceLog.kt:17]: I think this really convinces me that it should be called `DataStreamServiceLoggingProxy` :)

File: carp.common/src/commonMain/kotlin/dk/cachet/carp/common/infrastructure/services/ApplicationServiceRequest.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import kotlinx.serialization.KSerializer
+
+
+/**
+ * A request for [TService] stored in memory, which can be invoked using [invokeOn].
+ */
+interface ApplicationServiceRequest<TService, out TReturn>
+{
+    fun getResponseSerializer(): KSerializer<out TReturn>
+    suspend fun invokeOn( service: TService ): TReturn
```
Comment (ltj) [inline carp.common/src/commonMain/kotlin/dk/cachet/carp/common/infrastructure/services/ApplicationServiceRequest.kt:12]: Thank you. This seems more right and is easier to reason about than the previous one via constructor delegation 👍

File: carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/infrastructure/DeploymentServiceLog.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import dk.cachet.carp.deployments.application.users.ParticipantInvitation
+import dk.cachet.carp.protocols.application.StudyProtocolSnapshot
+import kotlinx.datetime.Instant
+
+
+/**
+ * A proxy for a deployment [service] which notifies of incoming requests and responses through [log]
+ * and keeps a history of requests in [loggedRequests].
+ */
+class DeploymentServiceLog(
```
Comment (ltj) [inline carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/infrastructure/DeploymentServiceLog.kt:20]: Naming

File: carp.studies.core/src/jvmTest/kotlin/dk/cachet/carp/studies/infrastructure/versioning/OutputRecruitmentServiceTestRequests.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import dk.cachet.carp.studies.application.StudyService
+import dk.cachet.carp.studies.infrastructure.RecruitmentServiceLog
+
+
+private val services = RecruitmentServiceHostTest.createService()
+
+class OutputProtocolServiceTestRequests :
+    OutputTestRequests<RecruitmentService>(
+        RecruitmentService::class,
+        RecruitmentServiceLog( services.first, services.third )
```
Comment (ltj) [inline carp.studies.core/src/jvmTest/kotlin/dk/cachet/carp/studies/infrastructure/versioning/OutputRecruitmentServiceTestRequests.kt:16]: Documentation is needed when accessing generic tuple indexes like this I think.

File: rpc/src/main/kotlin/dk/cachet/carp/rpc/Main.kt [hunk 5]
```diff
@@ -1,20 +1,33 @@
...
     // Create example requests for all request objects and responses of application service methods.
-    val exampleRequests = ApplicationServiceInfo.findInNamespace( "dk.cachet.carp" )
-        .associateWith { generateExampleRequests( it.serviceKlass, it.requestObjectKlass ) }
+    val exampleRequests = applicationServices
+        .associateWith { generateExampleRequests( it ) }
 
     // Output example requests.
-    FileUtils.deleteDirectory( JSON_EXAMPLES_FOLDER )
+    val jsonExamplesFolder = File( "build/rpc-examples/" )
+    FileUtils.deleteDirectory( jsonExamplesFolder )
```
Comment (ltj) [inline rpc/src/main/kotlin/dk/cachet/carp/rpc/Main.kt:27]: You probably want to catch `IOException` here

**Review by ltj — COMMENTED** [review]
Very nice work @Whathecode. Some general notes that may also have been addressed in my comments:
- General naming using the "Log" postfix. Think "LoggingProxy" is more appropriate
- (future) Concrete application service proxy implementations seems like something that can be generated. Vert.x service proxy codegen for inspiration https://vertx.io/docs/vertx-service-proxy/java/
- (future) Application service requests could potentially also be auto generated
- I noticed that the use of in-memory implementations is "hardcoded" everywhere. For shared test-sources we could consider making this pluggable, or otherwise just decide that we go for a test suite that replays the logged requests on any running instance of the infrastructure.

File: carp.studies.core/src/jvmTest/kotlin/dk/cachet/carp/studies/infrastructure/versioning/OutputRecruitmentServiceTestRequests.kt (hunk 4 above)
Comment (Whathecode) [reply carp.studies.core/src/jvmTest/kotlin/dk/cachet/carp/studies/infrastructure/versioning/OutputRecruitmentServiceTestRequests.kt]: Agreed. Replaced with data classes; a case where I do find self-documenting code is sufficient.

File: rpc/src/main/kotlin/dk/cachet/carp/rpc/Main.kt (hunk 5 above)
Comment (Whathecode) [reply rpc/src/main/kotlin/dk/cachet/carp/rpc/Main.kt:27]: I suspect this `rpc` application may become an entry point called as part of the Gradle build process to automate packaging test sources as part of releases.

If this fails, the build should fail. No clean way to handle the exception here for now. I think an exception by default results in an error code returned to the command line, which would be desired behavior. If the exception is unclear I can catch it to perhaps make it more meaningful, but will leave it as is for now.

**Pull request merged**